### ISSUE TITLE
Allow Prometheus to use ReadRequest_STREAMED_XOR_CHUNKS when calling remote read API.

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#5926]
https://github.com/prometheus/prometheus/pull/5703 extended remote read protocol and added `ReadRequest_STREAMED_XOR_CHUNKS`

Prometheus implements remote read API, but it can also be configured to **use/call** other remote read API. Enable streaming remote read if the server supports it.

AC:
* Prometheus remote read client is capable to use `ReadRequest_STREAMED_XOR_CHUNKS`

Similar client code for inspiration: https://github.com/thanos-io/thanos/pull/1268

**Comment by YaoZengzeng** [comment]
I'd like to fix this issue 😃 @bwplotka 

Maybe we can solve this problem by following steps:
1. Extend the `Read` method of `Client` in `storage/remote/client.go` to support both SAMPLE and STREAM, the return type changed to `*prompb.QueryResult` to `*http.Response` would be better.

2. Refactor `FromQueryResult` to decode both type of response and expose unified `storage.SeriesSet`.

3. Encapsulate the chunks into some struct to make it implement `storage.SeriesSet`. Actually the implementation method is similar to `chunkSeries` stuff in `tsdb/query.go`. In addition, this step may easy to be implemented if #5882 being merged :) 
( At first I‘d encapsulate the chunks to make the code run)

My initial idea is probably like this, any comments? @bwplotka

**Comment by bwplotka** [comment]
Sorry for lag and thanks @YaoZengzeng 

Yes that makes sense to me :+1: 

I think step 2 would be the most work, especially packing SAMPLED method to `SeriesSet`, but I guess we can't really avoid this.

For Streaming be careful as we might need to implement something like [this we did in Thanos](https://github.com/thanos-io/thanos/blob/master/pkg/query/iter.go#L39) in SeriesSet.

**Comment by YaoZengzeng** [comment]
@bwplotka Thanks for your comments and I'll fix this issue step by step 😃

**Comment by YaoZengzeng** [comment]
@bwplotka Please review #5974 again, otherwise the fixing of this issue can't move on 😂

**Comment by bwplotka** [comment]
Sorry for massive lag, proposed solution (: Also reach me on IRC bwplotka or [CNCF slack channel](https://slack.cncf.io/) bwplotka if you need to ping me - bit spammed on email.

**Comment by bwplotka** [comment]
Hi @YaoZengzeng this would be nice to have (: Any progress? What are the blockers? :hugs:

**Comment by Sudhar287** [comment]
Hello @YaoZengzeng. If you're not going to continue this, can I please take your work forward?

**Comment by YaoZengzeng** [comment]
@Sudhar287 Yes, Please :)

**Comment by Sudhar287** [comment]
Thank you for the/your quick response. :)

**Comment by bwplotka** [comment]
@Sudhar287 any update maybe? (:

**Comment by Sudhar287** [comment]
Hello @bwplotka! It seems that this PR #6042 cannot proceed until #6747 is completed according to [this](https://github.com/prometheus/prometheus/pull/6042#issuecomment-585953000). I was waiting for that to be finished first...Do you have a different proposal? :)

**Comment by bwplotka** [comment]
Wow, I think there is misunderstanding, I don't think anything is blocked and we can start working on it right way.

**Comment by bwplotka** [comment]
I commented https://github.com/prometheus/prometheus/pull/6042#issuecomment-619123029 

I don't think there was anything blocked... Sorry for confusion :facepalm: We are definitely good to go.

We kind of need this as we wrote on Thanos tool for serving CSV / JSON as some our API and we want to extend it to remote-read. Would be nice if we can chunks to Prometheus though :hugs:

**Comment by Sudhar287** [comment]
Thanks for the response and clarification. I'll start working on it! :)

**Comment by Sudhar287** [comment]
fyi, working on this actively. Will have the PR next week or so. :)

**Comment by bwplotka** [comment]
Awesome! (((:

**Comment by Tanemahuta** [comment]
Hello there, 

I have an implementation ready for which I refactored to ReadClient as follows:
```

**Comment by bwplotka** [comment]
It looks great @Tanemahuta, but we already have another PR https://github.com/prometheus/prometheus/pull/8351#pullrequestreview-628870312 

Maybe you can help us review it and maybe give your design as suggestions in the comments? 🤗

**Comment by Tanemahuta** [comment]
Will do. This changes the complete design and breaks the API though.

**Comment by detailyang** [comment]
@bwplotka any update? read STREAMED_XOR_CHUNKS response is important to reduce memory overhead and query latency:)

**Comment by hpple** [comment]
Hi, any updates on this issue?

**Comment by beorn7** [comment]
Hello from the bug scrub.

@bwplotka we believe the ball is in your court reviewing #11379 .

**Comment by bwplotka** [comment]
Thanks, yes - https://github.com/prometheus/prometheus/pull/11379 is done, closing this issue.

I don't think we have capacity to develop something on 2.x, but anyone is welcome to port this PR on 2.x on their own in their tmp fork until they can use 3.x (should be straightforward). 

What's stopping you from using 3.x? (let's continue this discussion on https://github.com/prometheus/prometheus/issues/12789)
